[Ready] Refactored diag interface, removed dummy sign.

davits opened the pull request:
# PR Prelude
- [x] I have read and understood YCM's [CONTRIBUTING][cont] document.
- [x] I have read and understood YCM's [CODE_OF_CONDUCT][code] document.
- [x] I have included tests for the changes in my PR. If not, I have included a
  rationale for why I haven't.
- [x] **I understand my PR may be closed if it becomes obvious I didn't
  actually perform all of these steps.**

# Why this change is necessary and useful

Removed dummy sign logic, now​ it places news signs, then removes obsolete ones. I did this in my fork long ago and never experienced any flicker.
Some source code refactoring, which I believe simplifies logic a little.

[cont]: https://github.com/Valloric/YouCompleteMe/blob/master/CONTRIBUTING.md
[code]: https://github.com/Valloric/YouCompleteMe/blob/master/CODE_OF_CONDUCT.md

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/valloric/youcompleteme/2677)
<!-- Reviewable:end -->

micbou commented:
Thanks for the PR.

> Removed dummy sign logic, not it places news signs, then removes obsolete ones. I did this in my fork long ago and never experienced any flicker.

I wonder why it was not done this way in the first place.

---

Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[python/ycm/diagnostic_interface.py, line 123 at r1](https://reviewable.io:443/reviews/valloric/youcompleteme/2677#-Kltje8PmxqYNB-R0oIm:-Kltje8PmxqYNB-R0oIn:bunim9j) ([raw file](https://github.com/valloric/youcompleteme/blob/22be4e777d64cd6af1fc3ae3cc6d79501135bfb7/python/ycm/diagnostic_interface.py#L123)):*
> ```Python
> 
>   def _UpdateSquiggles( self ):
>     if self._bufnr != vim.current.buffer.number:
> ```

We could use `GetCurrentBufferNumber` here.

---

*[python/ycm/diagnostic_interface.py, line 144 at r1](https://reviewable.io:443/reviews/valloric/youcompleteme/2677#-KlyaGmt7375kwApo04_:-KlyaGmt7375kwApo04a:b-2fjfjs) ([raw file](https://github.com/valloric/youcompleteme/blob/b20809332c3298cb954cff20006cb52955b54b2d/python/ycm/diagnostic_interface.py#L144)):*
> ```Python
>   for diags in itervalues( line_to_diags ):
>     # Insert squiggles in reverse order so that errors overlap warnings.
>     for diag in reversed( diags ):
> ```

Why did you change that? We want error squiggles to be drawn over warning ones, not the other way around.

---

*[python/ycm/diagnostic_interface.py, line 205 at r1](https://reviewable.io:443/reviews/valloric/youcompleteme/2677#-Kltg6j27rltiN9Xbqkn:-Kltg6j27rltiN9Xbqko:b-nwxeb2) ([raw file](https://github.com/valloric/youcompleteme/blob/22be4e777d64cd6af1fc3ae3cc6d79501135bfb7/python/ycm/diagnostic_interface.py#L205)):*
> ```Python
>     for diags in itervalues( self._line_to_diags ):
>       # We also want errors to be listed before warnings so that errors aren't
>       # hidden by the warnings; Vim won't place a sign oven an existing one.
> ```

`over` instead of `oven`.
 

---

*[python/ycm/tests/test_utils.py, line 142 at r1](https://reviewable.io:443/reviews/valloric/youcompleteme/2677#-KltkS-Qu8ADONq9QHBU:-KltkS-Qu8ADONq9QHBV:bloictm) ([raw file](https://github.com/valloric/youcompleteme/blob/22be4e777d64cd6af1fc3ae3cc6d79501135bfb7/python/ycm/tests/test_utils.py#L142)):*
> ```Python
> def _MockVimMatchEval( value ):
>   if value == 'getmatches()':
>     return list( VIM_MATCHES )
> ```

It's already supposed to be a list.

---

*[python/ycm/tests/test_utils.py, line 277 at r1](https://reviewable.io:443/reviews/valloric/youcompleteme/2677#-KlthXTLBA2Ce-DGp4KV:-KlthXTLBA2Ce-DGp4KW:b-x07cps) ([raw file](https://github.com/valloric/youcompleteme/blob/22be4e777d64cd6af1fc3ae3cc6d79501135bfb7/python/ycm/tests/test_utils.py#L277)):*
> ```Python
>       return self.group
>     elif key == 'id':
>       return self.id
> ```

Are you sure this is needed? `group` and `id` are properties of this class.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2677#-:-Klthf4RliENWv9IYJ8G:b2vcbb8)*
<!-- Sent from Reviewable.io -->

davits commented:
Here is the dummy sign change https://github.com/Valloric/YouCompleteMe/commit/3fcc63127f58840bcc362a0a7b7b2f359c25db4e

It seems that the problem was, if there were no new signs placed, removing obsolete signs one by one were causing flickering. It was ~3 years ago, must be fixed by now.
Can you confirm that it is gone ?

---

Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[python/ycm/diagnostic_interface.py, line 123 at r1](https://reviewable.io:443/reviews/valloric/youcompleteme/2677#-Kltje8PmxqYNB-R0oIm:-KlymJyOizOUX5hSDL-n:b-kl2oyo) ([raw file](https://github.com/valloric/youcompleteme/blob/22be4e777d64cd6af1fc3ae3cc6d79501135bfb7/python/ycm/diagnostic_interface.py#L123)):*
<details><summary><i>Previously, micbou wrote…</i></summary><blockquote>

We could use `GetCurrentBufferNumber` here.
</blockquote></details>

Also I need to remember why I added this check and make it into a comment :)
Maybe I thought, that since vim matches are window specific, it is better be safe then sorry.

---

*[python/ycm/diagnostic_interface.py, line 144 at r1](https://reviewable.io:443/reviews/valloric/youcompleteme/2677#-KlyaGmt7375kwApo04_:-Klykv7L3_K_f3UETavC:bg0tm9k) ([raw file](https://github.com/valloric/youcompleteme/blob/b20809332c3298cb954cff20006cb52955b54b2d/python/ycm/diagnostic_interface.py#L144)):*
<details><summary><i>Previously, micbou wrote…</i></summary><blockquote>

Why did you change that? We want error squiggles to be drawn over warning ones, not the other way around.
</blockquote></details>

My bad.

---

*[python/ycm/tests/test_utils.py, line 142 at r1](https://reviewable.io:443/reviews/valloric/youcompleteme/2677#-KltkS-Qu8ADONq9QHBU:-Klygxo5yyNYYM-MFivA:bsz95ix) ([raw file](https://github.com/valloric/youcompleteme/blob/22be4e777d64cd6af1fc3ae3cc6d79501135bfb7/python/ycm/tests/test_utils.py#L142)):*
<details><summary><i>Previously, micbou wrote…</i></summary><blockquote>

It's already supposed to be a list.
</blockquote></details>

Returning a copy, because `ClearYcmSyntaxMatches()` iterates over the returned list and removes matches from VIM_MATCHES.
Otherwise it iterates over and removes from the same list, and is working incorrectly.

---

*[python/ycm/tests/test_utils.py, line 277 at r1](https://reviewable.io:443/reviews/valloric/youcompleteme/2677#-KlthXTLBA2Ce-DGp4KV:-KlygpyKXJSTXNJM3ViP:b-cftkig) ([raw file](https://github.com/valloric/youcompleteme/blob/22be4e777d64cd6af1fc3ae3cc6d79501135bfb7/python/ycm/tests/test_utils.py#L277)):*
<details><summary><i>Previously, micbou wrote…</i></summary><blockquote>

Are you sure this is needed? `group` and `id` are properties of this class.
</blockquote></details>

It is required from the `vimsupport.ClearYcmSyntaxMatches()`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2677#-:-Klyno6nNCNPOwpfOfAy:be8kol5)*
<!-- Sent from Reviewable.io -->

Valloric commented:
I roughly remember the original change; the flicker was a real problem. I could be that it's largely resolved today because of changes in Vim. If so, we can remove the hack and just tell people to update to latest Vim.

If folks have tried out this patch and there's no flicker, I'm fine with it.

---

Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[python/ycm/tests/test_utils.py, line 142 at r1](https://reviewable.io:443/reviews/valloric/youcompleteme/2677#-KltkS-Qu8ADONq9QHBU:-Km-9a3QGbZS4cAyOo70:bkqoyby) ([raw file](https://github.com/valloric/youcompleteme/blob/22be4e777d64cd6af1fc3ae3cc6d79501135bfb7/python/ycm/tests/test_utils.py#L142)):*
<details><summary><i>Previously, davits (Davit Samvelyan) wrote…</i></summary><blockquote>

Returning a copy, because `ClearYcmSyntaxMatches()` iterates over the returned list and removes matches from VIM_MATCHES.
Otherwise it iterates over and removes from the same list, and is working incorrectly.
</blockquote></details>

Then there should probably be a comment explaining this.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2677#-:-Km-8Cf0SDgVxLrhE96T:b-s0h0o7)*
<!-- Sent from Reviewable.io -->

davits commented:
Addressed concerns raised during review.
Found and fixed bug when `YouCompleteMe._current_buffer` were not updating correctly/in time. It was happening because `FileType` event in Vim comes before `BufEnter`, and reparse from `BufEnter` is blocked by  `s:VisitedBufferRequiresReparse()`. So it will end up sending file parse event for the previous buffer from `FileType` and updating current buffer to current one on `BufVisit`, but not sending parse event for it.
I've added current buffer set in `FileType` as a quick solution, but in long term I want to revisit `s:VisitedBufferRequiresReparse()` logic.

---

Review status: 2 of 6 files reviewed at latest revision, 5 unresolved discussions.

---

*[python/ycm/diagnostic_interface.py, line 123 at r1](https://reviewable.io:443/reviews/valloric/youcompleteme/2677#-Kltje8PmxqYNB-R0oIm:-KmFN4T81ECKkTmBDIPm:bss4a2b) ([raw file](https://github.com/valloric/youcompleteme/blob/22be4e777d64cd6af1fc3ae3cc6d79501135bfb7/python/ycm/diagnostic_interface.py#L123)):*
<details><summary><i>Previously, davits (Davit Samvelyan) wrote…</i></summary><blockquote>

Also I need to remember why I added this check and make it into a comment :)
Maybe I thought, that since vim matches are window specific, it is better be safe then sorry.
</blockquote></details>

Removed.

---

*[python/ycm/diagnostic_interface.py, line 144 at r1](https://reviewable.io:443/reviews/valloric/youcompleteme/2677#-KlyaGmt7375kwApo04_:-KmFM3TroPNFmx4Kxmiv:b-896fix) ([raw file](https://github.com/valloric/youcompleteme/blob/b20809332c3298cb954cff20006cb52955b54b2d/python/ycm/diagnostic_interface.py#L144)):*
<details><summary><i>Previously, davits (Davit Samvelyan) wrote…</i></summary><blockquote>

My bad.
</blockquote></details>

Done.

---

*[python/ycm/diagnostic_interface.py, line 205 at r1](https://reviewable.io:443/reviews/valloric/youcompleteme/2677#-Kltg6j27rltiN9Xbqkn:-KmFJr0QtO-hb6hAbSoP:b-896fix) ([raw file](https://github.com/valloric/youcompleteme/blob/22be4e777d64cd6af1fc3ae3cc6d79501135bfb7/python/ycm/diagnostic_interface.py#L205)):*
<details><summary><i>Previously, micbou wrote…</i></summary><blockquote>

`over` instead of `oven`.
 

</blockquote></details>

Done.

---

*[python/ycm/tests/test_utils.py, line 142 at r1](https://reviewable.io:443/reviews/valloric/youcompleteme/2677#-KltkS-Qu8ADONq9QHBU:-KmFN-IsK4jR9E9fCEuf:b-896fix) ([raw file](https://github.com/valloric/youcompleteme/blob/22be4e777d64cd6af1fc3ae3cc6d79501135bfb7/python/ycm/tests/test_utils.py#L142)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary><blockquote>

Then there should probably be a comment explaining this.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2677#-:-KmFZkH0LypSZ58yjGOB:b-3ry35g)*
<!-- Sent from Reviewable.io -->

davits commented:
The case I described can be reproduced by:
vim file1.cpp
:e file2.cpp
If second file has errors they don't show up.

---

Review status: 2 of 6 files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2677#-:-KmGTT4cAx1mpJw2yzz-:bp24roz)*
<!-- Sent from Reviewable.io -->

micbou commented:
I think we should drop the current buffer caching. I profiled the `SetCurrentBuffer` method and it takes less than a microsecond to execute.

---

Reviewed 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[python/ycm/diagnostic_interface.py, line 197 at r2](https://reviewable.io:443/reviews/valloric/youcompleteme/2677#-KmH-Bg9b9TZz9PmC8lR:-KmH-Bg9b9TZz9PmC8lS:b9lwkze) ([raw file](https://github.com/valloric/youcompleteme/blob/e5b0565d1a804e6383e4b4c54150b5cdd263238d/python/ycm/diagnostic_interface.py#L197)):*
> ```Python
>       bufnr = vimsupport.GetBufferNumberForFilename( location[ 'filepath' ] )
>       if bufnr != self._bufnr:
>           continue
> ```

Indentation should be two spaces, not four.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2677#-:-KmH3-w-e53Svf_2XzOn:b2jznld)*
<!-- Sent from Reviewable.io -->

davits commented:
@micbou did you mean `_GetCurrentBuffer` from initial iteration of my previous commit ?

---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2677#-:-KmHzS_RN_Ukob3sjU7V:b-jl7uiy)*
<!-- Sent from Reviewable.io -->

micbou commented:
Yes, though I didn't profile `_GetCurrentBuffer` since it was removed. Both methods should take the same amount of time.

---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2677#-:-KmI7muvX4XSbk1FgvfW:b-li066t)*
<!-- Sent from Reviewable.io -->

davits commented:
Yeah, the implementation with `_GetCurrentBuffer` wouldn't have such problem.
I am going to convert back to it.

---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2677#-:-KmIYug2HxmTAxELhFlm:b-9192wx)*
<!-- Sent from Reviewable.io -->

davits commented:
Review status: 2 of 8 files reviewed at latest revision, 1 unresolved discussion.

---

*[python/ycm/diagnostic_interface.py, line 197 at r2](https://reviewable.io:443/reviews/valloric/youcompleteme/2677#-KmH-Bg9b9TZz9PmC8lR:-KmR1zz9QauIi8TDaoaA:b-896fix) ([raw file](https://github.com/valloric/youcompleteme/blob/e5b0565d1a804e6383e4b4c54150b5cdd263238d/python/ycm/diagnostic_interface.py#L197)):*
<details><summary><i>Previously, micbou wrote…</i></summary><blockquote>

Indentation should be two spaces, not four.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2677)*
<!-- Sent from Reviewable.io -->

micbou commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Just one thing. As mentioned in issue https://github.com/Valloric/YouCompleteMe/issues/2680, I think we should remove deleted buffers from the `_buffers` dictionary on the `BufferUnload` event.

---

Reviewed 6 of 6 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2677#-:-KmRCehmoaS-fwHYdi2R:bud3iuv)*
<!-- Sent from Reviewable.io -->

davits commented:
@micbou, I did a little investigation:
During the session, the same file is always loaded with the same buffer number, and it isn't important whether buffer was unloaded, or deleted (closed and removed from buffer list).

The thing that troubles me most is the fact that the buffer retains its sign bar even if  unloaded, and by removing unloaded buffer from `_buffers` list we loose the information about that signs. 
This is a problem in a following situation: user renames some function in a cpp file and gets an error, because there is no matching declaration in a header file, user switches to the header file, renames function there as well, and switches back to the cpp file, the error is resolved, but the sign remains.

Of course we can clear signs on the buffer unload as well. But having signs disappear after :e other_file, :b# is kind of a "buggy" behavior.

On the other side, keeping growing list of `_buffers` isn't going to affect performance, even for a thousands of buffers.

I am in favor of keeping the buffers.

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2677#-:-KmW994C8Q52jmr2f8sd:b1fo5lv)*
<!-- Sent from Reviewable.io -->

micbou commented:
> During the session, the same file is always loaded with the same buffer number, and it isn't important whether buffer was unloaded, or deleted (closed and removed from buffer list).

That's good to know. I am fine with keeping the buffers then.

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2677#-:-KmWOCvSoOHqtpoy9gc4:b-ceci18)*
<!-- Sent from Reviewable.io -->

davits commented:
:thumbsup:  marking the PR as Ready then.

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2677#-:-KmWTR8aBTn8Mb2sjuoV:b-djwykj)*
<!-- Sent from Reviewable.io -->

micbou commented:
I couldn't reproduce the flickers in console and graphical Vim so I think it's good to go.

@zzbot r+

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2677#-:-KmXFx2aZlIQWF_65ivt:bd0ohrp)*
<!-- Sent from Reviewable.io -->
